File: frontend/src/modules/orcamentos/utils/investimentoOferta.ts
import type { LinhaEditavelOrcamento } from '../types/orcamentoLinha'
import type {
  OrcamentoConfiguradorPainelDto,
  OrcamentoPreviewOfertaItemDto,
  PerfilOferta,
} from '../types/orcamentos'
import {
  INVESTIMENTO_DESCRICAO_DEMAIS_PADRAO,
  descricaoInvestimentoConsolidadaPadrao,
  descricaoInvestimentoExibicao,
} from './investimentoDescricao'
import { normalizarNcmInvestimento } from './ncmInvestimento'
import { parseDecimalPt } from './orcamentoUi'

function refPainel(painel: OrcamentoConfiguradorPainelDto): string {
  return `P${painel.ordem + 1}`
}

function decimalStr(n: number): string {
  if (!Number.isFinite(n)) return '0'
  const arred = Math.round(n * 100) / 100
  return String(arred)
}

function subtotalLinha(linha: LinhaEditavelOrcamento): number {
  const qtd = parseDecimalPt(linha.quantidade || '0')
  const preco = parseDecimalPt(linha.preco_unitario || '0')
  if (!Number.isFinite(qtd) || !Number.isFinite(preco)) return 0
  return qtd * preco
}

function linhaInvestimento(
  descricao: string,
  subtotal: number,
  ncm: string
): OrcamentoPreviewOfertaItemDto {
  const valor = decimalStr(subtotal)
  return {
    descricao,
    quantidade: '1',
    preco_unitario: valor,
    subtotal: valor,
    unidade: 'un',
    ncm,
  }
}

function investimentoPorPainel(
  linhas: LinhaEditavelOrcamento[],
  paineis: OrcamentoConfiguradorPainelDto[],
  ncm: string,
  investimentoDescricao?: string
): OrcamentoPreviewOfertaItemDto[] | null {
  const ativos = [...paineis]
    .filter((p) => p.modo === 'ATIVO')
    .sort((a, b) => a.ordem - b.ordem)
  if (!ativos.length) return null

  const itens: OrcamentoPreviewOfertaItemDto[] = []
  const usados = new Set<number>()

  for (const painel of ativos) {
    const ref = refPainel(painel)
    const grupo = linhas.filter((linha, idx) => {
      if (!ref || !linha.painelRef) return false
      if (linha.painelRef === ref) {
        usados.add(idx)
        return true
      }
      return false
    })
    if (!grupo.length) continue
    const sub = grupo.reduce((s, l) => s + subtotalLinha(l), 0)
    itens.push(linhaInvestimento(painel.descricao_painel.trim() || 'Painel', sub, ncm))
  }

  const semPainel = linhas.filter((_, idx) => !usados.has(idx))
  if (semPainel.length) {
    const sub = semPainel.reduce((s, l) => s + subtotalLinha(l), 0)
    itens.push(
      linhaInvestimento(
        descricaoInvestimentoExibicao(
          investimentoDescricao,
          INVESTIMENTO_DESCRICAO_DEMAIS_PADRAO
        ),
        sub,
        ncm
      )
    )
  }

  return itens.length ? itens : null
}

export function montarItensInvestimentoLocal(params: {
  perfil: PerfilOferta
  titulo: string
  linhas: LinhaEditavelOrcamento[]
  paineis?: OrcamentoConfiguradorPainelDto[]
  ncmInvestimento?: string
  investimentoDescricao?: string
}): {
  modo: 'ITENS_UNITARIOS' | 'CONSOLIDADO' | 'POR_PAINEL'
  itens: OrcamentoPreviewOfertaItemDto[]
} {
  const { perfil, titulo, linhas, paineis = [], ncmInvestimento, investimentoDescricao } =
    params

  if (perfil === 'SOLUCAO_COMPLETA') {
    const ncm = normalizarNcmInvestimento(ncmInvestimento)
    const porPainel = investimentoPorPainel(linhas, paineis, ncm, investimentoDescricao)
    if (porPainel) {
      return { modo: 'POR_PAINEL', itens: porPainel }
    }
    const total = linhas.reduce((s, l) => s + subtotalLinha(l), 0)
    return {
      modo: 'CONSOLIDADO',
      itens: [
        linhaInvestimento(
          descricaoInvestimentoExibicao(
            investimentoDescricao,
            descricaoInvestimentoConsolidadaPadrao(titulo)
          ),
          total,
          ncm
        ),
      ],
    }
  }

  return {
    modo: 'ITENS_UNITARIOS',
    itens: linhas.map((linha, index) => ({
      id: linha.id || `linha-${index}`,
      codigo: linha.produtoCodigo || linha.servicoCodigo || '',
      descricao: linha.descricao || '—',
      quantidade: linha.quantidade || '0',
      preco_unitario: linha.preco_unitario || '0',
      subtotal: decimalStr(subtotalLinha(linha)),
      unidade: linha.servicoUnidadeMedida?.trim() || 'un',
      ncm: linha.tipo === 'PRODUTO' ? linha.produtoNcm?.trim() || '' : '',
    })),
  }
}

File: backend/apps/orcamentos/services/investimento_oferta.py
"""Montagem da tabela de investimento da oferta (itens unitários, painéis ou consolidado)."""
from __future__ import annotations

from decimal import Decimal

from apps.orcamentos.models import (
    ModoConfiguradorPainelChoices,
    Orcamento,
    OrcamentoItem,
    PerfilOfertaChoices,
)
from apps.orcamentos.services.investimento_descricao import (
    INVESTIMENTO_DESCRICAO_DEMAIS_PADRAO,
    descricao_investimento_consolidada_padrao,
    descricao_investimento_exibicao,
)
from apps.orcamentos.services.ncm_investimento import ncm_investimento_orcamento


def _decimal_str(value: Decimal | None) -> str:
    if value is None:
        return "0"
    normalized = value.normalize()
    if normalized == normalized.to_integral():
        return str(normalized.quantize(Decimal("1")))
    return format(normalized, "f")


def _subtotal(item: OrcamentoItem) -> Decimal:
    return item.quantidade * item.preco_unitario


def _item_linha(item: OrcamentoItem) -> dict:
    return {
        "id": str(item.id),
        "ordem": item.ordem,
        "tipo": item.tipo,
        "codigo": item.servico.codigo
        if item.servico_id
        else item.produto.codigo
        if item.produto_id
        else "",
        "descricao": item.descricao,
        "quantidade": _decimal_str(item.quantidade),
        "preco_unitario": _decimal_str(item.preco_unitario),
        "subtotal": _decimal_str(_subtotal(item)),
        "unidade": item.servico.unidade_medida if item.servico_id else "",
        "ncm": item.produto.ncm if item.produto_id and item.produto.ncm else "",
    }


def _linha_investimento(
    *,
    descricao: str,
    subtotal: Decimal,
    ncm: str = "",
    unidade: str = "un",
) -> dict:
    return {
        "descricao": descricao,
        "quantidade": "1",
        "preco_unitario": _decimal_str(subtotal),
        "subtotal": _decimal_str(subtotal),
        "unidade": unidade,
        "ncm": ncm,
    }


def _investimento_materiais(itens: list[OrcamentoItem]) -> dict:
    return {
        "modo": "ITENS_UNITARIOS",
        "titulo": "Investimento",
        "itens": [_item_linha(item) for item in itens],
    }


def _investimento_por_painel(
    orcamento: Orcamento,
    itens: list[OrcamentoItem],
    ncm: str,
    *,
    descricao_demais_padrao: str = INVESTIMENTO_DESCRICAO_DEMAIS_PADRAO,
) -> dict | None:
    paineis = list(
        orcamento.configuradores_painel.filter(
            modo=ModoConfiguradorPainelChoices.ATIVO,
        ).order_by("ordem", "id")
    )
    if not paineis:
        return None

    itens_por_painel: dict = {p.id: [] for p in paineis}
    sem_painel: list[OrcamentoItem] = []
    for item in itens:
        if item.configurador_painel_id and item.configurador_painel_id in itens_por_painel:
            itens_por_painel[item.configurador_painel_id].append(item)
        else:
            sem_painel.append(item)

    linhas: list[dict] = []
    for painel in paineis:
        grupo = itens_por_painel.get(painel.id) or []
        if not grupo:
            continue
        sub = sum((_subtotal(i) for i in grupo), Decimal("0"))
        linhas.append(
            _linha_investimento(
                descricao=painel.descricao_painel.strip() or "Painel",
                subtotal=sub,
                ncm=ncm,
            )
        )

    if sem_painel:
        sub = sum((_subtotal(i) for i in sem_painel), Decimal("0"))
        linhas.append(
            _linha_investimento(
                descricao=descricao_investimento_exibicao(
                    orcamento, descricao_demais_padrao
                ),
                subtotal=sub,
                ncm=ncm,
            )
        )

    if not linhas:
        return None

    return {
        "modo": "POR_PAINEL",
        "titulo": "Investimento",
        "itens": linhas,
    }


def _investimento_solucao_completa(orcamento: Orcamento, itens: list[OrcamentoItem]) -> dict:
    ncm = ncm_investimento_orcamento(orcamento)
    por_painel = _investimento_por_painel(orcamento, itens, ncm)
    if por_painel:
        return por_painel

    total = sum((_subtotal(item) for item in itens), Decimal("0"))
    return {
        "modo": "CONSOLIDADO",
        "titulo": "Investimento",
        "itens": [
            _linha_investimento(
                descricao=descricao_investimento_exibicao(
                    orcamento,
                    descricao_investimento_consolidada_padrao(orcamento.titulo),
                ),
                subtotal=total,
                ncm=ncm,
            )
        ],
    }


def montar_investimento_oferta(orcamento: Orcamento, itens: list[OrcamentoItem]) -> dict:
    perfil = orcamento.perfil_oferta or PerfilOfertaChoices.MATERIAIS
    if perfil == PerfilOfertaChoices.SOLUCAO_COMPLETA:
        return _investimento_solucao_completa(orcamento, itens)
    return _investimento_materiais(itens)
